### ISSUE TITLE
Auto-assign milestones to merged PR's

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse-platform/eclipse.platform.releng.aggregator#2203]
Follow up on https://github.com/eclipse-jdt/eclipse.jdt.core/discussions/2627

We should automatically assign milestones to merged PR's to simplify our life tracking which release contained which fix etc.

There are various solutions possible, see list below fro what google found:

- https://github.com/eclipse-theia/theia/pull/8631
- https://github.com/DataDog/dd-trace-java/blob/master/.github/workflows/add-milestone-to-pull-requests.yaml
- https://www.jessesquires.com/blog/2022/08/04/automatically-assign-milestones-with-github-actions/

This should be done for all eclipse SDK projects (Platform, JDT, PDE, Equinox).

Note, differently to Platform, JDT has BETA_JAVA23 milestone](https://github.com/eclipse-jdt/eclipse.jdt.core/milestones?direction=asc&sort=due_date&state=open) that should be excluded from the "automated" process.

Beside this I believe all projects should be very similar in what milestones we have, as they are created from a job during release preparation process AFAIK.

@HannesWell, @laeubi : you seem to have some experience in the github actions area, I would really appreciate if you could help here.

**Comment by HannesWell** [comment]
Sounds good and is probably not too difficult to implement (haven't looked into the examples yet).
I can try to help with that next week (I'm away over the weekend).

**Comment by laeubi** [comment]
For Tycho I use `is:pr is:merged no:milestone` and then batch assign milestones or `is:issue no:milestone is:closed`.

> We should automatically assign milestones to merged PR's to simplify our life tracking which release contained which fix etc.

Github also allows to see what has changed when one creates a release and can generate a list or one can generate it form the Tags created in a repository.

Beside that assigning a milestone to a *closed* PR is somewhat useless, a Milestones main purpose is to track what works needs to be done (or is planned) see for example https://github.com/eclipse-tycho/tycho/milestones but from my experience no one is really doing such planning, but your BETA_23 is a good example, if there is work that needs to be done until one can declare "finish" of 23 support then a milestone (e.g. Java 23 Support) would be a good choice to let people know what is left.

**Comment by iloveeclipse** [comment]
> Beside that assigning a milestone to a _closed_ PR is somewhat useless

There are JDT & platform committers that don't think so, me including. Also looking on the links, there are projects like theia which are also doing this, so it seem to be not useless also for other projects.

I have almost daily task to identify whether some fix or feature or bug exists / doesn't exit in a specific release. Currently this means to lookup commit in a repo, scroll up/down the history to see in which release / milestone it is included. It is tedious and wastes my time.

> Milestones main purpose is to track what works needs to be done

Sure, it is one purpose, and important one. Unfortunately it is hard to "enforce" every contributor to follow every good practice, and people also do mistakes. Since this is a different topic, if you are interested, please open a discussion thread for the planning /milestones use etc. 

Let's keep discussion on this ticket focused on the one concrete use case - automatically track which PR is included in which milestone.

**Comment by BeckerWdf** [comment]
> > Beside that assigning a milestone to a _closed_ PR is somewhat useless
> 
> There are JDT & platform committers that don't think so, me including. Also looking on the links, there are projects like theia which are also doing this, so it seem to be not useless also for other projects.
> 
> I have almost daily task to identify whether some fix or feature or bug exists / doesn't exit in a specific release. Currently this means to lookup commit in a repo, scroll up/down the history to see in which release / milestone it is included. It is tedious and wastes my time.
> 
> > Milestones main purpose is to track what works needs to be done
> 
> Sure, it is one purpose, and important one. Unfortunately it is hard to "enforce" every contributor to follow every good practice, and people also do mistakes. Since this is a different topic, if you are interested, please open a discussion thread for the planning /milestones use etc.
> 
> Let's keep discussion on this ticket focused on the one concrete use case - automatically track which PR is included in which milestone.

I agree with Andrey. Milestone can be used in two use-cases. Planning (when do want to deliver that PR) and book-keeping (in which release this change was made).

Setting the milestone (if not yet done) after merging sounds reasonable to me.

**Comment by jukzi** [comment]
+1 autoset milestone after merge to master
